### PR TITLE
Composite Page with MVC

### PR DESCRIPTION
**Pull request by AhmedObaidi**
I refactored the application so pagelets will have their own real MVC structure, which means that each pagelet will have controller responsible for making the Rest call and rendering the content. 

The main page template will only call the sub-templates for providing place holders. And the main controller will only call the sub-controllers for providing HtmlStream.

I would really appreciate any comment, since I'm just starting with both Scala and Play.

### CODE REVIEW

File: app/controllers/Mock.scala [hunk 1]
```diff
@@ -14,9 +14,9 @@ object Mock extends Controller {
 
   def mock(serviceName: String) = Action.async {
     serviceName match {
-      case "wvyp" => respond(data = "56", delay = 10)
-      case "search" => respond(data = "10", delay = 3000)
-      case "likes" => respond(data = "150", delay = 40)
+      case "wvyp" => respond(data = "56", delay = 3000)
```
Comment (brikis98) [inline app/controllers/Mock.scala]: Probably no need to commit the changes to this file.

File: app/controllers/Aggregator.scala [hunk 2]
```diff
@@ -1,26 +0,0 @@
-package controllers
-
-import play.api.mvc.{Action, Controller}
-import play.api.libs.concurrent.Execution.Implicits._
-import ui.Pagelet
-
-/**
- * This controller shows an example of composing together the results of two other controllers togheter: Wvyp and Wvyu.
- */
-object Aggregator extends Controller {
```
Comment (brikis98) [inline app/controllers/Aggregator.scala]: Why delete this controller?

File: app/controllers/EnumeratorExamples.scala [hunk 3]
```diff
@@ -1,31 +0,0 @@
-package controllers
-
-import play.api.mvc.Controller
-import play.api.mvc.Action
-import play.api.libs.iteratee.Enumerator
-import play.api.libs.concurrent.Promise
-import play.api.libs.concurrent.Execution.Implicits._
-
-object EnumeratorExamples extends Controller {
```
Comment (brikis98) [inline app/controllers/EnumeratorExamples.scala]: Why delete this controller? 

To provide some context, this is a repo that I use as part of my talk on Composable and Streamable Play Apps ([slides](http://www.slideshare.net/brikis98/composable-and-streamable-play-apps), [video](https://www.youtube.com/watch?v=4b1XLka0UIw)), and this controller and many of the others are a part of that talk.

File: app/controllers/ProfileViewsStreamController.scala [hunk 4]
```diff
@@ -1,27 +1,25 @@
...
 /**
  * An example standalone module that represents the "Who's Viewed Your Profile" module from the LinkedIn homepage.
- * It's identical to the Wvyp controller, except that the data is streamed to the browser as soon as it's available in
+ * It's identical to the ProfileViewsController controller, except that the data is streamed to the browser as soon as it's available in
  * small "pagelets", similar to Facebook's BigPipe.
  */
-object WvypStream extends Controller {
+object ProfileViewsStreamController {
+
+  def stream = HtmlStream {
```
Comment (brikis98) [inline app/controllers/ProfileViewsStreamController.scala]: If it's not an `Action`, the class does not extend `Controller`, and it doesn't show up in `routes`, it is not really a controller. That defeats the whole "standalone endpoint" idea.

File: app/controllers/UpdateViewsStreamController.scala [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+import ui.{HtmlStream, Pagelet}
+
+/**
+ * An example standalone module that represents the "Who's Viewed Your Profile" module from the LinkedIn homepage.
+ * It's identical to the ProfileViewsController controller, except that the data is streamed to the browser as soon as it's available in
+ * small "pagelets", similar to Facebook's BigPipe.
+ */
+object UpdateViewsStreamController {
+
+  def stream = HtmlStream {
```
Comment (brikis98) [inline app/controllers/UpdateViewsStreamController.scala]: Same comment about not being a controller anymore.

File: app/controllers/Wvyp.scala [hunk 6]
```diff
@@ -1,24 +0,0 @@
-package controllers
-
-import play.api.mvc.{Cookie, Action, Controller}
-import data.ServiceClient
-import play.api.libs.concurrent.Execution.Implicits._
-
-/**
- * An example standalone module that represents the "Who's Viewed Your Profile" module from the LinkedIn homepage.
- */
-object Wvyp extends Controller {
```
Comment (brikis98) [inline app/controllers/Wvyp.scala]: Same question about why delete?

File: app/controllers/WvypEnumerator.scala [hunk 7]
```diff
@@ -1,22 +0,0 @@
-package controllers
-
-import play.api.mvc.Controller
-import play.api.mvc.Action
-import data.ServiceClient
-import play.api.libs.iteratee.Enumerator
-import play.api.libs.concurrent.Execution.Implicits._
-
-object WvypEnumerator extends Controller {
```
Comment (brikis98) [inline app/controllers/WvypEnumerator.scala]: Same question about why delete?

File: app/controllers/WvypStreamJava.java [hunk 8]
```diff
@@ -1,45 +0,0 @@
...
-import play.libs.F;
-import play.mvc.Controller;
-import play.mvc.Result;
-import scala.Function1;
-import ui.HtmlStream;
-import ui.Pagelet;
-
-import java.util.Arrays;
-
-public class WvypStreamJava extends Controller
```
Comment (brikis98) [inline app/controllers/WvypStreamJava.java]: Same question about why delete? This is a Java example to show that these techniques work in Java as well as Scala.

File: app/controllers/Wvyu.scala [hunk 9]
```diff
@@ -1,24 +0,0 @@
-package controllers
-
-import play.api.mvc.{Action, Controller}
-import data.ServiceClient
-import play.api.libs.concurrent.Execution.Implicits._
-
-/**
- * An example standalone module that represents the "Who's Viewed Your Updates" module from the LinkedIn homepage.
- */
-object Wvyu extends Controller {
```
Comment (brikis98) [inline app/controllers/Wvyu.scala]: Same question about why delete?

**Comment by brikis98** [comment]
I left a few comments in the code, but to be honest, I'm not sure I fully understand the goal of the pull request. Can you describe it a bit more? Thanks!

File: app/controllers/ProfileViewsStreamController.scala (hunk 4 above)
Comment (AhmedObaidi) [inline app/controllers/ProfileViewsStreamController.scala]: Good point, I changed to be Controller now and can be served as a page
```scala
object UpdateViewsStreamController  extends ChunkedController {

  override def stream = HtmlStream {

    val likesCountFuture = ServiceClient.makeServiceCall("likes")
    val commentsCountFuture = ServiceClient.makeServiceCall("comments")

    val likesStream =Pagelet.renderStream(likesCountFuture.map(str => views.html.pagelets.updates.likesCount(str.toInt)), "likeCount")
    HtmlStream.interleave(likesStream)
  }

  override def holder(): HtmlStream = views.stream.pagelets.updates.updates()
}
```

My concern was, and correct me if I'm wrong, that the composed page was doing the REST calls and not calling the pagelet controller, so we ended up writing 2 controllers instead.

File: app/controllers/EnumeratorExamples.scala (hunk 3 above)
Comment (AhmedObaidi) [inline app/controllers/EnumeratorExamples.scala]: Sorry, I was just deleting these files so I can focus on the concept itself. Brilliant video btw!

**Comment by brikis98** [comment]
Haven't heard back in a few months. Closing.

**Pull request closed without merging**